lib/logstash/inputs/google_pubsub.rb

```diff
@@ -142,140 +172,62 @@ class LogStash::Inputs::GooglePubSub < LogStash::Inputs::Base
-        :api_method => @pubsub.projects.subscriptions.create,
-        :parameters => {'name' => @subscription},
-        :body_object => {
-          :topic => @topic,
-          :ackDeadlineSeconds => 15
-        }
-      )
-      if result.error? and result.status != 409
-        raise Google::APIClient::ClientError, "Error #{result.status}: #{result.error_message}"
+      @logger.debug("Creating subscription #{@subscription_id}")
+      subscriptionAdminClient = SubscriptionAdminClient.create
+      begin
+        subscriptionAdminClient.createSubscription(@subscription_name, @topic_name, PushConfig.getDefaultInstance(), 0)
+      rescue
+        @logger.info("Subscription already exists")
       end
       @subscription_exists = true
-    end # if !@subscription
-
-    @logger.debug("Pulling messages from sub '#{subscription}'")
-    while !stop?
-      # Pull and queue messages
-      messages = []
-      result = request(
-        :api_method => @pubsub.projects.subscriptions.pull,
-        :parameters => {'subscription' => @subscription},
-        :body_object => {
-          :returnImmediately => false,
-          :maxMessages => @max_messages
-        }
-      )
+    end
 
-      if !result.error?
-        messages = JSON.parse(result.body)
-        if messages.key?("receivedMessages")
-          messages = messages["receivedMessages"]
-        end
-      else
-        @logger.error("Error pulling messages:'#{result.error_message}'")
+    @logger.debug("Pulling messages from sub '#{@subscription_id}'")
+    handler = MessageReceiver.new do |message|
+      # handle incoming message, then ack/nack the received message
+      data = message.getData().toStringUtf8()
+      @codec.decode(data) do |event|
+        event.set("host", event.get("host") || @host)
+        decorate(event)
+        queue << event
       end
-
-      if messages
-        messages.each do |msg|
-          if msg.key?("message") and msg["message"].key?("data")
-            decoded_msg = Base64.decode64(msg["message"]["data"])
-            begin
-              parsed_msg = JSON.parse(decoded_msg)
-            rescue
-              parsed_msg = { :raw_message => decoded_msg }
-            end
-            event = LogStash::Event.new(parsed_msg)
-            decorate(event)
-            queue << event
-          end
-        end
-
-        ack_ids = messages.map{ |msg| msg["ackId"] }
-        result = request(
-          :api_method => @pubsub.projects.subscriptions.acknowledge,
-          :parameters => {'subscription' => @subscription},
-          :body_object => {
-            :ackIds => ack_ids
-          }
-        )
-        if result.error?
-          @logger.error("Error #{result.status}: #{result.error_message}")
-        end
-      end # if messages
-    end # loop
-  end # def run
-end # class LogStash::Inputs::GooglePubSub
+    end
+    listener = SubscriberListener.new do |from, failure|
+      @logger.error("#{failure}")
+      exit
```

Comment:
I'm not sure if exit is the right thing to do here, but I'm not super familiar with Ruby.

Reply:
I think that's a good idea, that way Logstash can report and handle it appropriately. I think Logstash tries to restart the plugin on failure which might cause a loop, but if the failure was because of a temporary fault it's probably the best option.